r1cs/gadgets/std/src/bits/macros.rs

```diff
@@ -0,0 +1,2532 @@
+                            None => ConstraintF::one(), // delta is 0, so we can set any value
+                        }
+                    });
+
+                    let inv_var = FpGadget::<ConstraintF>::alloc(cs.ns(|| "alloc inv"), || inv.ok_or(SynthesisError::AssignmentMissing))?;
+
+                    // enforce constraints:
+                    // (1 - is_smaller) * delta_lc = 0 enforces that is_smaller == 1 when delta != 0, i.e., when a < b
+                    // inv * delta_lc = is_smaller enforces that is_smaller == 0 when delta == 0, i.e., when b >= a
+                    cs.enforce(|| "enforce is smaller == true", |_| is_smaller.not().lc(CS::one(), ConstraintF::one()), |lc| lc + &delta_lc, |lc| lc);
+                    cs.enforce(|| "enforce is smaller == false", |lc| &inv_var.get_variable() + lc, |lc| lc + &delta_lc, |_| is_smaller.lc(CS::one(), ConstraintF::one()));
+
+                    Ok(is_smaller)
+                }
+
+                fn enforce_smaller_than<CS: ConstraintSystemAbstract<ConstraintF>>
+                (&self, mut cs: CS, other: &Self) -> Result<(), SynthesisError> {
+                    // first enforce that self <= other, which holds iff other - self does not underflow
+                    let diff = {
+                        let mut multi_eq = MultiEq::new(&mut cs);
+                        other.sub_noborrow(&mut multi_eq, self)?
+                    };
+                    // then, enforce that other - self is non zero, which holds iff the difference
+                    // has at least a non zero bit
+                    Boolean::enforce_or(cs.ns(|| "enforce self != other"), &diff.bits)
+                }
+
+                // override the default implementation to exploit the fact that enforcing constraint
+                // is cheaper than computing a Boolean gadget with the comparison outcome
+                fn enforce_cmp<CS: ConstraintSystemAbstract<ConstraintF>>(
+                    &self,
+                    mut cs: CS,
+                    other: &Self,
+                    ordering: Ordering,
+                    should_also_check_equality: bool,
+                ) -> Result<(), SynthesisError> {
+                    let (left, right) = match (ordering, should_also_check_equality) {
+                        (Ordering::Less, false) | (Ordering::Greater, true) => (self, other),
+                        (Ordering::Greater, false) | (Ordering::Less, true) => (other, self),
+                        (Ordering::Equal, _) => return self.enforce_equal(cs, other),
+                    };
+
+                    if should_also_check_equality {
+                        left.enforce_greater_or_equal_than(cs.ns(|| "enforce greater equal"), right)
+                    } else {
+                        left.enforce_smaller_than(cs.ns(|| "enforce smaller than"), right)
+                    }
+                }
+
+            }
+
+
+
+            #[cfg(test)]
+            mod test {
```

Comment:
I saw in many test performing equality assertions using the "enforce_equal" gadget. It's a bit overdoing (unless it's the test for the EqGadget itself lol) and slows down the tests, it's more than enough comparing the gadget with primitive values (like you do in the test_uint_gadget_value) function.